Sam/Fix/Numb Keys

samholmes opened the pull request:
Re-implement `derivePublicKey` to follow `edge-core-js` key-formats spec

- Add PrivateKey and PublicKey types and asPrivateKey and asPublicKey
cleaners for internal use and documentation.
- Desensitize public keys returned by derivePublicKey using the cleaners.
- Modify util functions to always include segwit and wrapped-segwit
formats together, following the spec.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1201256727099457

samholmes commented on src/common/plugin/makeCurrencyTools.ts:
```diff
@@ -51,15 +51,18 @@ export function makeCurrencyTools(
-      const key = utxoUtils.getXpubKey({ coin: engineInfo.network })
-      // TODO: which xpub should be saved? the root path (m) or hardened path with the wallet format path (m/{purpose}'/{coinType}'/{account}')?
-      const publicKey: CurrencyFormatKeys = utxoUtils.deriveXpubsFromKeys({
+      const privateKey = asPrivateKey(
```
Define an `asCurrencyPrivateKey` alias cleaner in the outerscope.

samholmes commented on src/common/plugin/types.ts:
```diff
@@ -17,7 +17,8 @@ export enum NetworkEnum {
 
-export type CurrencyFormat = 'bip32' | 'bip44' | 'bip49' | 'bip84'
+export type CurrencyFormat = ReturnType<typeof asCurrencyFormat>
+export const asCurrencyFormat = asValue('bip32', 'bip44', 'bip49', 'bip84')
```
Discover if we can remove `bip84`.

samholmes replied:
I don't think we can safely remove this from the union type because this type appears to be a general purpose union of all formats.

samholmes reviewed (commented):
Remove all the utils cruft in favor of a simple object map that maps the private key formats to the public key formats based on the key-formats standard. The map is then reduced to the public key format. Don't use `if` statements; this PR should be -200 lines diff.

samholmes commented:
Feedback from discussions with @swansontec : 

* Remove `WalletKeys` and don't pass along `walletKeys` (which includes private keys) throughout the plugin. Instead only pass along the `PublicKey` and `PrivateKey` objects where needed. We should avoid passing along the sensitive data that is in either `walletInfo` and `walletKeys`.
* Change `PublicKey` type from a union of format records to a single record where the formats are optional.

swansontec commented on src/common/utxobased/keymanager/cleaners.ts:
```diff
@@ -0,0 +1,164 @@
+        coin: coinName,
+        network: networkType ?? NetworkEnum.Mainnet
+      })
+      return asPublicKey({
```
This data has a known data type, so it should not need to be cleaned. Please replace the `asPublicKey` with a type delaration on the IIFE return value, so Typescript checks it instead of the cleaner.

swansontec commented on src/common/utxobased/engine/utils.ts:
```diff
@@ -202,12 +165,12 @@ export const deriveXprivFromKeys = (args: {
 }): CurrencyFormatKeys => {
   const xpubs: CurrencyFormatKeys = {}
-  for (const format of getWalletSupportedFormats(args)) {
+  for (const format of asWalletKeysFormats(args.privateKey)) {
```
`args.privateKey` is typed, and therefore should *not* be cleaned. Instead, do `getSupportedFormats(privateKey.format)`.

swansontec commented on src/common/utxobased/keymanager/cleaners.ts:
```diff
@@ -0,0 +1,164 @@
+  })
+})
+
+/**
+ * This cleaner returns the wallet's supported formats as specified in the
+ * key-formats specification.
+ */
+export const asWalletKeysFormats: Cleaner<CurrencyFormat[]> = (
+  keys: any
+): CurrencyFormat[] => {
+  const format = asOptionalCurrencyFormat(keys?.format)
+
+  switch (format) {
+    case 'bip32':
+      return ['bip32']
+    case 'bip44':
+      return ['bip44']
+    case 'bip49':
+      return ['bip49', 'bip84']
+    case 'bip84':
+      return ['bip49', 'bip84']
+    default:
+      throw new Error(`Unsupported format ${format}`)
+  }
+}
```
This should not be a cleaner - it is being used in contexts where data has already been cleaned, so it should just be a normal utility function taking `CurrencyFormat` and returning `CurrencyFormat[]`. Let's call it `getSupportedFormats`, for the sake of this PR.

swansontec commented on src/common/utxobased/keymanager/cleaners.ts:
```diff
@@ -0,0 +1,164 @@
+        publicKeys: currencyFormatKeys
+      })
+    })()
+
```
Given the problems with `asWalletKeysFormats`, change to:

```
const format = asOptionalCurrencyFormat(keys.format)
return {
  ...,
  keys: {
    format,
    supportedFormats: getSupportedFormats(format),
    ...
```

swansontec commented on src/common/plugin/makeCurrencyTools.ts:
```diff
@@ -30,36 +33,32 @@ export function makeCurrencyTools(
 
-      return {
-        ...keys,
+      const privateKey = asCurrencyPrivateKey({
```
Use a type definition here, instead of a cleaner:

```
const privateKey: PrivateKey = {
  privateKey: mnemonic,
  format: opts?.format ?? engineInfo.formats?.[0] ?? 'bip44',
  coinType: opts?.coinType ?? engineInfo.coinType ?? 0
}
```

swansontec commented on src/common/utxobased/engine/makeUtxoEngine.ts:
```diff
@@ -177,14 +189,15 @@ export async function makeUtxoEngine(
-      return getMnemonic({ keys: walletInfo.keys, coin: engineInfo.network })
+      // Private key may be missing for watch-only wallets
+      const privateKey = asMaybe(asCurrencyPrivateKey)(sensitiveWalletInfo.keys)
+      if (privateKey == null)
```
I think the intention was to return `null` here when unavailable, rather than throwing.

swansontec reviewed (changes requested):
A few more changes are needed, to avoid double-cleaning data.

samholmes commented:
Squashed the first two commits as suggested (rebasing the two commits into smaller changes was too time consuming). I added fixup commits to hopefully make it clearer that all the feedback has been taken into consideration.

swansontec commented on src/common/utxobased/keymanager/cleaners.ts:
```diff
@@ -0,0 +1,173 @@
+
+    // Prepare fields to return for NumbWalletInfo:
+    const { id, type } = walletInfo
+    const coinType = asCoinType(walletInfo.keys.coinType)
+    const format = asOptionalCurrencyFormat(walletInfo.keys.format)
+    const supportedFormats = getSupportedFormats(format)
```
I didn't notice this before, but this logic won't work in watch-only mode. It assumes we have a private key, which isn't a given.

I think this function could be split into two sections like:

```js
const private = asMaybe(asCurrencyPrivateKey)(walletInfo.keys)
if (private != null) {
  // convert private to public
  return { format: private.format, ... }
}

const public = asMaybe(asCurrencyPublicKey)(walletInfo.keys)
if (public != null) {
  const formats = Object.keys(public.publicKeys).filter(asWalletFormat)
  return { format: formats[0], supportedFormats: formats, ... }
}

// If we get here, nothing matched:
throw new Error('either a public or private key are required')
```

This structure gets rid of the IIFE, and since we just try the keys both ways an go with whichever works.

This would allow `asNumbWalletInfo` to work in watch-only mode. Right now it will just blow up.

swansontec reviewed (changes requested):
More changes needed.

samholmes commented on src/common/utxobased/engine/makeUtxoEngine.ts:
```diff
@@ -158,12 +167,12 @@ export async function makeUtxoEngine(
       return {
-        walletId: walletInfo.id.split(' - ')[0],
-        walletType: walletInfo.type,
+        walletId: sensitiveWalletInfo.id.split(' - ')[0],
+        walletType: sensitiveWalletInfo.type,
```
Use the numbed type

samholmes commented on src/common/utxobased/engine/makeUtxoEngineState.ts:
```diff
@@ -92,13 +91,15 @@ export function makeUtxoEngineState(
     network,
     pluginInfo,
-    walletInfo,
+    numbWalletInfo: walletInfo,
```
Change it to just `walletInfo`

thehobbit85 commented:
/rebase-autosquash

samholmes commented on src/common/utxobased/keymanager/cleaners.ts:
```diff
@@ -0,0 +1,189 @@
+
+    const publicKey = asMaybe(asPublicKey)(walletInfo.keys)
+    if (publicKey != null) {
+      const formats = Object.entries(publicKey.publicKeys)
+        // Filter out undefined values in the entries because cleaners allow
+        // undefined values for optional fields.
+        .filter(([, value]) => value != null)
+        // Map to the entry's key as a currency format
+        .map(([format]) => asMaybe(asCurrencyFormat)(format))
+        // Filter out any formats that didn't pass the cleaner assertion
+        .filter(
+          (format?: CurrencyFormat): format is CurrencyFormat => format != null
+        )
```
This had to be fixed. The formats needed to be changed from accessing the `Object.keys(publicKey.publicKeys)` to `Object.entries(publicKey.publicKeys)` so that way we can filter out the values (because they may be undefined).
